chirho/robust/internals/linearize.py

```diff
@@ -0,0 +1,158 @@
+    rdotr = torch.dot(r, r)
+    v = rdotr / torch.dot(p, z)
+    newrdotr = rdotr
+    mu = newrdotr / rdotr
+
+    zeros_xr = torch.zeros_like(x)
+
+    for _ in range(cg_iters):
+        not_converged = rdotr > residual_tol
+        z = torch.where(not_converged, f_Ax(p), z)
+        v = torch.where(not_converged, rdotr / torch.dot(p, z), v)
+        x += torch.where(not_converged, v * p, zeros_xr)
+        r -= torch.where(not_converged, v * z, zeros_xr)
+        newrdotr = torch.where(not_converged, torch.dot(r, r), newrdotr)
+        mu = torch.where(not_converged, newrdotr / rdotr, mu)
+        p = torch.where(not_converged, r + mu * p, p)
+        rdotr = torch.where(not_converged, newrdotr, rdotr)
+
+        # rdotr = newrdotr
+        # if rdotr < residual_tol:
+        #     break
+    return x
+
+
+def conjugate_gradient_solve(f_Ax: Callable[[T], T], b: T, **kwargs) -> T:
+    flatten, unflatten = make_flatten_unflatten(b)
+
+    def f_Ax_flat(v: torch.Tensor) -> torch.Tensor:
+        v_unflattened: T = unflatten(v)
+        result_unflattened = f_Ax(v_unflattened)
+        return flatten(result_unflattened)
+
+    return unflatten(_flat_conjugate_gradient_solve(f_Ax_flat, flatten(b), **kwargs))
+
+
+def make_empirical_fisher_vp(
+    func_log_prob: Callable[Concatenate[ParamDict, Point[T], P], torch.Tensor],
+    log_prob_params: ParamDict,
+    data: Point[T],
+    *args: P.args,
+    **kwargs: P.kwargs,
+) -> Callable[[ParamDict], ParamDict]:
+    batched_func_log_prob: Callable[[ParamDict, Point[T]], torch.Tensor] = torch.vmap(
+        lambda p, data: func_log_prob(p, data, *args, **kwargs),
+        in_dims=(None, 0),
+        randomness="different",
+    )
+
+    def bound_batched_func_log_prob(params: ParamDict) -> torch.Tensor:
+        return batched_func_log_prob(params, data)
+
+    def jvp_fn(v: ParamDict) -> torch.Tensor:
+        return torch.func.jvp(bound_batched_func_log_prob, (log_prob_params,), (v,))[1]
+
+    vjp_fn = torch.func.vjp(bound_batched_func_log_prob, log_prob_params)[1]
+
+    def _empirical_fisher_vp(v: ParamDict) -> ParamDict:
+        jvp_log_prob_v = jvp_fn(v)
+        return vjp_fn(jvp_log_prob_v / jvp_log_prob_v.shape[0])[0]
+
+    return _empirical_fisher_vp
+
+
+def linearize(
+    model: Callable[P, Any],
+    guide: Callable[P, Any],
+    *,
+    num_samples_outer: int,
+    num_samples_inner: Optional[int] = None,
+    max_plate_nesting: Optional[int] = None,
+    cg_iters: Optional[int] = None,
+    residual_tol: float = 1e-10,
+) -> Callable[Concatenate[Point[T], P], ParamDict]:
+    assert isinstance(model, torch.nn.Module)
+    assert isinstance(guide, torch.nn.Module)
+    if num_samples_inner is None:
+        num_samples_inner = num_samples_outer**2
+
+    predictive = pyro.infer.Predictive(
+        model,
+        guide=guide,
+        num_samples=num_samples_outer,
+        parallel=True,
+    )
+    predictive_params, func_predictive = make_functional_call(predictive)
```

Comment:
Why do we need to use `make_functional_call` here? `func_predictive` is used in L150, but we don't require gradients with respect to `data` anyways

Reply:
I believe it's necessary for `vmap` to compose correctly with `linearize`.